src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -0,0 +1,396 @@
+        }
+        return futureResult;
+    }
+
+    void shutdown()
+    {
+        checkExecutor.shutdown();
+        compactionRequests.values().forEach(promises -> promises.forEach(p -> p.cancel(false)));
+    }
+
+    int getOngoingCompactionsCount(ColumnFamilyStore cfs)
+    {
+        return ongoingCompactions.count(cfs);
+    }
+
+    @Override
+    public void run()
+    {
+        logger.trace("Running background compactions check");
+
+        // We shuffle the CFSs for which the compaction was requested so that with each run we traverse those CFSs
+        // in different order and make each CFS have equal chance to be selected
+        ArrayList<ColumnFamilyStore> compactionRequestsList = new ArrayList<>(compactionRequests.keySet());
+        Collections.shuffle(compactionRequestsList, random);
+
+        for (ColumnFamilyStore cfs : compactionRequestsList)
+        {
+            List<SettableFuture<RequestResult>> promises = compactionRequests.remove(cfs);
+
+            logger.trace("Checking compaction request for {}.{} - there are {} awaiting promises",
+                         cfs.getKeyspaceName(),
+                         cfs.getTableName(),
+                         promises.size());
+
+            RequestResult completionStatus = canRunCompactionTasks(cfs, promises);
+            if (completionStatus != null)
+            {
+                promises.stream().filter(p -> !p.isDone()).forEach(p -> p.set(completionStatus));
+                continue;
+            }
+
+            logger.trace("Scheduling a background task check for {}.{} with {}",
+                         cfs.keyspace.getName(),
+                         cfs.name,
+                         cfs.getCompactionStrategy().getName());
+
+            Supplier<CompletableFuture<?>> compactionTask = Optional.ofNullable(lazyRunCompactionTasks(cfs))
+                                                                    .orElseGet(() -> lazyRunUpgradeTasks(cfs));
+
+            if (compactionTask != null)
+            {
+                ongoingCompactions.add(cfs);
+
+                compactionTask.get().handle((ignored, throwable) -> {
+                    ongoingCompactions.remove(cfs);
+
+                    if (throwable != null)
+                    {
+                        logger.warn(String.format("Aborting compaction of %s.%s due to error", cfs.getKeyspaceName(), cfs.getTableName()), throwable);
+                        handleCompactionError(throwable, cfs);
+
+                        promises.forEach(p -> p.setException(throwable));
+                    }
+                    else
+                    {
+                        logger.trace("Finished compaction for {}.{}", cfs.getKeyspaceName(), cfs.getTableName());
+
+                        // Since we have run at least one task and thus the set of sstables has changed,
+                        // check for new compaction possibilities
+                        requestCompaction(cfs);
```

Comment:
This code will not request when any task fails (`allOf`), but the request actually needs to be made if any of the tasks succeeded. Since the latter is not easily checkable at this point, we should request on both error and success (and copy the comment that explains this).

Reply:
Got it. Though I'll try to implement it in the way it retries in at least one compaction completes